api/src/main/java/org/apache/iceberg/Table.java

```diff
@@ -373,4 +374,14 @@ default Snapshot snapshot(String name) {
 
     return null;
   }
+
+  /**
+   * Returns the statistics file for the given snapshot id, if available.
+   *
+   * @return the {@link StatisticsFile} for the given snapshot id, if available.
+   */
+  default Optional<StatisticsFile> statistics(long snapshotId) {
```

Comment:
Thanks for adding this, I think this API is in the right direction but I wonder if it's taking enough information to be able to resolve the right file. There's an implicit assumption in the current implementation of this API that a Puffin file will have all the blobs needed. I think it's possible that there could multiple puffins, and one puffin has blob type NDV and the other puffin has blob type SomeFutureIndex. both of these puffin files are for snapshot 1.

In the current implementation if someone wants the Puffin where there's SomeFutureIndex, but we happen to return the Puffin due to the generic "Find me a puffin with the given snapshot logic" I feel like the API isn't really doing what it needs.

Here's the signature I'm thinking at the moment:

default StatisticsFile statisticsFor(long snapshotId, String blobType)
This will attempt to attempt to find the statistics file produced at the given snapshot which contains the blob type.

I also think that in case a statistics file for exactly that snapshot couldn't be found, we should return the latest statistics rather than just return nothing. While it's possible that the statistics are not completely accurate in this approach, I think in the average case data distributions wouldn't change so drastically between snapshots that the statistics would work horribly against the query. It's probably better in the average case to have some statistics. If there's no there's no statistics file containing the desired blob then I think we'd just return null.

WDYT?

Reply:
yes @amogh-jahagirdar your suggestion is perfect, considering a generic solution where we support multiple bolb types. The current implementation is considering that we will only support the "apache-datasketches-theta-v1".
We recently faced this when we were dealing with presto, considering both engines were using a common catalog, and hence the puffin file created by presto was not use-able as it was of a different blob type "presto-sum-data-size-bytes-v1". This change would be a more of a futuristic change which we may take up.

Regarding the best effort search of stats @amogh-jahagirdar, I thing we need to reconsider if we want to have some statistics always, because that would depend on the amount of data added or deleted after the last time we ran and Analyze. Because stale statistics could lead to wrong query plans. And what if we let the user configure ho much deviation or change is the user fine with to continue using the older statistics. For the same I had made some changes so that the user may decide the amount of change https://github.com/karuppayya/iceberg/compare/fix_snapshot...jeesou:fix_snapshot_modifications?expand=1.

Kindly have a look at it @amogh-jahagirdar and @karuppayya and share your suggestions please.
I have not considered the delete scenario, if i find any deletion happening I am not using old stats, but that can be up to discussion as delete is a tricky subject in this case.